lib/elixir/src/elixir_expand.erl

```diff
@@ -272,6 +272,22 @@ expand({for, Meta, [_ | _] = Args}, E) ->
         form_error(Meta, ?key(E, file), ?MODULE, {missing_option, for, [do]})
     end,
 
+  IsReducing = lists:keymember(reduce, 1, Opts),
+  IsClauses = is_list(Expr) andalso lists:any(fun({'->', _, _}) -> true; (_) -> false end, Expr),
```

Comment:
Should this check `lists:all` and not `any`?

Reply:
It is always all or none.  So theoretically we need to check only the first.